docs/build_parameters.rst

```diff
@@ -0,0 +1,833 @@
+            # Remove ImageChange triggers
+            # Do not handle plugins
+
+        # Override
+        def render_name(self):
+            # Re-implement to use BuildUserParams
+
+        # Override
+        def render_node_selectors(self):
+            # Re-implement to use BuildUserParams
+
+        # Override
+        def render(self):
+            # Validate BuildUserParams
+
+            # Render name
+            # Render resource limits
+
+            # Set template.spec.source.git.uri
+            # Set template.spec.source.git.ref
+
+            # Set template.spec.output.to.name
+
+            # Set template.spec.triggers[0].imageChange.from.name
+
+            # Set template.spec.strategy.customStrategy.from.kind
+            # Set template.spec.strategy.customStrategy.from.name
+
+            # Set git-repo-name label
+            # Set git-branch label
+            # Set koji-task-id label
+
+            # Set template.spec.strategy.customStrategy.env[] USER_PARAMS
+
+            # Adjust for repo info
+            # Adjust for scratch build
+            # Adjust for isolated build
+            # Adjust for custom base image
+
+            # Set required_secrets based on reactor_config
+            # Set worker_token_secrets based on reactor_config, if any
+
+            # Log build json
+            # Return build json
+
+
+The new class ``BuildUserParams`` will be added. This class will be similar to
+``BuildSpec`` class, but will handle a much smaller set of parameters. It should
+also provide a method to convert it to and from json.
+
+
+Rendering Plugins
+"""""""""""""""""
+
+Once the build is started control is handed over to atomic-reactor. Its input
+plugin ``osv3`` is responsible for loading the plugin configuration from the
+environment variable **ATOMIC_REACTOR_PLUGINS**. If this environment variable is
+not found the plugin will look for the environment variable **USER_PARAMS**. If
+found, a new code path will generate the plugin configuration on the fly.
+
+The new osbs-client method ``render_plugins_configuration`` will generate the
+plugin configuration based on the value of **USER_PARAMS**. As previously
+mentioned, environment configuration will be retrieved as needed by each
+atomic-reactor plugin. The generated plugin configuration will contain the order
+in which plugins will run as well as user parameters.
+
+Support for environment variable **DOCK_PLUGINS** will be removed from ``osv3``.
+
+``render_plugins_configuration`` pseudo code in osbs/api.py::
+
+    class OSBS(object):
+        def render_plugins_configuration(self, user_params):
+            user_params = BuildUserParams.from_json(user_params)
+
+            return PluginsConfigurationRender(user_params).render()
+
+The new class ``PluginsConfigurationRender`` will be responsible for actually
+rendering each plugin. Some of its logic will be taken from ``BuildRequest``,
+and ``DockJsonManipulator``.
```

Comment:
Duplicated, or can any be shared? (Can be sorted out during development probably)

Reply:
`DockJsonManipulater` has some logic to explicitly handle `ATOMIC_REACTOR_PLUGINS` environment variable. It might be cleaner to have its own implementation, but as you said this might be more obvious during development, yes.

Reply:
Added sentence:
> Whether functionality of ``DockJsonManipulater`` is duplicated or reused will be clearer during implementation.